### ISSUE TITLE
[UX] Views: Field rewriting: See if the limitation of having to add a field *before* the one to be rewritten can be eliminated.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2605]
See how in views, if you want to use another field's value when rewriting the output of another field, in order for that field's value to be made available, you have to place that field **before** the field that you are rewriting the output for? (text bolded for emphasis):

![image](https://cloud.githubusercontent.com/assets/2423362/24519807/f41435ba-15d1-11e7-8911-25cf6933aaff.png)

...why is that? Can we somehow remove that limitation?

And while at it, can we also please change the requirement to be using `%5B` or `%5D` with something easier like `[[` `]]` instead?

**Comment by olafgrabienski** [comment]
> you have to place that field before the field that you are rewriting ...why is that?

Guessing: to make the field technically available for the following one.

To remove this limitation would be a 'nice to have' for me, as field rewriting is however quite advanced.

**Comment by markabur** [comment]
I agree that the "before" requirement is a little bit of a head-scratcher, but it does help prevent circular field rewrites. E.g. without the requirement you would be able to put the [title] token in the [type] field and vice-versa, which is an infinite loop.

**Comment by klonos** [comment]
@markabur re the infinite loop: hmm, I guess that this can happen if these tokens output the rewritten value. If they output the actual raw value of the fields, then it is a non issue. Will have to locate and check the code...

**Comment by klonos** [comment]
This still seems silly to me. I'll start a series of small tests to see how this whole thing works...

**Comment by klonos** [comment]
Test environment:

- a "3fields" content type with 3 text (short) fields: field1, field2, and field3
- a "3fields" node that has these values:
-- field1: "123"
-- field2: "abc"
-- field3: "!@#"
- a view of "3fields" content, showing fields in this order: field1, field2, field3
-- field1 rendered as is
-- field2 rewritten as `[field_field1] [field_field2]`
-- field3 rewritten as `[field_field2] [field_field3]`

Result:
-- field1: 123
-- field2: 123 abc
-- field3: 123 abc !@#

So, `[field_field2]` within the context of field2 is its own value, while within the context of field3 is is rewritten value. Hmm. What if I wanted to render the non-rewritten value of field2 in field3 in order to achieve this result instead:

-- field1: 123
-- field2: 123 abc
-- field3: abc !@#

So, field3 gives me the following available tokens:

- [field_field1] == Content: field1
- [field_field2] == Content: field2
- [field_field3] == Content: field3
- [field_field3-value] == Raw value
- [field_field3-format] == Raw format

A couple of issues there:

1. Why can't I have a `[field_field2-value]` (and a `[field_field1-value]` for that matter)?
2. (off topic to this issue, but still) why are these tokens using `-` instead of `:`? ...for consistency with the path tokens, I'd expect `[field_field3:value]`

**Comment by klonos** [comment]
So it comes down to this... if with the above setup I need to get:

- field1: 123
- field2: 123 abc
- field3: abc !@#

instead of:

- field1: 123
- field2: 123 abc
- field3: 123 abc !@#

I then need to:

- add field2 *twice*; once rewritten as `[field_field1] [field_field2]`, and once without rewriting
- exclude the second instance of field2 from display (it gets `field2_1` as machine name)
- rewrite field3 as `[field_field2_1] [field_field3]` instead of `[field_field2] [field_field3]`

Why have users go through all these steps?? ...that is if they figure it out.

**Comment by klonos** [comment]
...on to testing if I can create infinite loops :smile:

**Comment by klonos** [comment]
>I agree that the "before" requirement is a little bit of a head-scratcher, but it does help prevent circular field rewrites. E.g. without the requirement you would be able to put the [title] token in the [type] field and vice-versa, which is an infinite loop.

I think that bottom line is that if the tokens render the rewritten values of the fields, then yes there could be an infinite loop. But if the tokens render the raw value of the fields (as what's in the database), then there would not be such an issue.

**Comment by klonos** [comment]
...and a bit more testing (and confusion added):

With the above-mentioned setup:

- add a "html" text (long) field to the "3fields" content type (I know it ain't 3 fields any more 😄 ) and set it to allow filtered HTML. Save the changes.
- edit the test node and add some rich text to the html field (nothing fancy; some bold/italics and a link will do). Save the node.
- back to the view, add the html field 3 times:
    - 1st instance: change its label to "html content", rewrite the results as `[field_html]`
    - 2nd instance: change the label to "html raw value", rewrite the results as `[field_html_1-value]`
    - 3rd instance: change the label to "html raw format", rewrite the results as `[field_html_2-format]`

Result:

AFAICT, there is no difference between the normal value of the field and its raw value. I tried to find documentation for this, but the terms "drupal views rewrite raw value" seem to be too generic. Any one have a clue?
